ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/GetSyncBlocksFromPeerTask.java

```diff
@@ -0,0 +1,154 @@
+import org.hyperledger.besu.ethereum.eth.manager.PendingPeerRequest;
+import org.hyperledger.besu.ethereum.eth.messages.BlockBodiesMessage;
+import org.hyperledger.besu.ethereum.eth.messages.EthPV62;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Requests bodies from a peer by header, matches up headers to bodies, and returns blocks. */
+public class GetSyncBlocksFromPeerTask extends AbstractPeerRequestTask<List<SyncBlock>> {
+  private static final Logger LOG = LoggerFactory.getLogger(GetSyncBlocksFromPeerTask.class);
+
+  private final List<BlockHeader> headers;
+  private final Map<BodyIdentifier, List<BlockHeader>> bodyToHeaders = new HashMap<>();
+  private final ProtocolSchedule protocolSchedule;
+
+  private GetSyncBlocksFromPeerTask(
+      final EthContext ethContext,
+      final List<BlockHeader> headers,
+      final MetricsSystem metricsSystem,
+      final ProtocolSchedule protocolSchedule) {
+    super(ethContext, EthPV62.GET_BLOCK_BODIES, metricsSystem);
+    checkArgument(!headers.isEmpty());
+
+    this.protocolSchedule = protocolSchedule;
+    this.headers = headers;
+    headers.forEach(
+        (header) -> {
+          final BodyIdentifier bodyId = new BodyIdentifier(header);
+          bodyToHeaders.putIfAbsent(bodyId, new ArrayList<>(headers.size()));
+          bodyToHeaders.get(bodyId).add(header);
+        });
+  }
+
+  public static GetSyncBlocksFromPeerTask forHeaders(
+      final EthContext ethContext,
+      final List<BlockHeader> headers,
+      final MetricsSystem metricsSystem,
+      final ProtocolSchedule protocolSchedule) {
+    return new GetSyncBlocksFromPeerTask(ethContext, headers, metricsSystem, protocolSchedule);
+  }
+
+  @Override
+  protected PendingPeerRequest sendRequest() {
+    final List<Hash> blockHashes =
+        headers.stream().map(BlockHeader::getHash).collect(Collectors.toList());
+    LOG.atTrace()
+        .setMessage("Requesting {} bodies with hashes {}.")
+        .addArgument(blockHashes.size())
+        .addArgument(blockHashes)
+        .log();
+    final long minimumRequiredBlockNumber = headers.getLast().getNumber();
+
+    return sendRequestToPeer(
+        peer -> {
+          LOG.atTrace()
+              .setMessage("Requesting {} bodies from peer {}.")
+              .addArgument(blockHashes.size())
+              .addArgument(peer)
+              .log();
+          return peer.getBodies(blockHashes);
+        },
+        minimumRequiredBlockNumber);
+  }
+
+  @Override
+  protected Optional<List<SyncBlock>> processResponse(
```

Comment:
Could the same class be used for SyncBlocks and Blocks? Looks like the only place they differ in implementation is in the processResponse

Reply:
Done